tetramer_validator/validate.py

```diff
@@ -11,147 +11,332 @@
 PTM_file = "data/PTM_list.tsv"
 PTM_file = path.join(here, PTM_file)
 
+errors = []
 with open(PTM_file) as fh_1:
     reader = csv.DictReader(fh_1, delimiter="\t")
     PTM_display = [name["display_name"] for name in reader]
 
 
-def validate(pep_seq, mhc_name=None, mod_type=None, mod_pos=None):
+def generate_problem_table(level, rule_name, value, instructions, fix=None):
+    global errors
+    errors.append(locals())
+    print("generae")
+    print(errors)
+
+def validate(pep_seq, mhc_name, mod_type=None, mod_pos=None):
+    """Main validate function."""
     args = locals()
-    null_str = "NULL"
-    if null_str in args.values():
-        return "NULL value entered. If there is no need for a particular value, please leave blank."
-    if "" == args["pep_seq"]:
-        return "Incorrect number of arguments: Peptide sequence is required"
-    if pep_seq != "" and mhc_name is None and mod_type is None and mod_pos is None:
-        return (
-            "Incorrect number of arguments: Please enter modification information "
-            "and/or MHC molecule information."
+    global errors
+    errors = []
+    properNumArguments(args)
+    null_input_check(args)
+
+    if errors:
+        return errors
+
+    if mod_pos and mod_type:
+        validate_peptide(pep_seq, mod_type=mod_type, mod_pos=mod_pos)
+    print(errors)
+    validate_mhc_name(mhc_name)
+    print("validate")
+    print(errors)
+    return errors
+
+
+def properNumArguments(args):
+    """Checks for proper number and combination of arguments for validate function"""
+
+    incorrect_num_str = "IncorrectNumArgs"
+
+    if not args["pep_seq"]:
+        generate_problem_table(
+            level="error",
+            rule_name=incorrect_num_str,
+            value=args["pep_seq"],
+            instructions="Enter peptide sequence",
         )
-    # Thanks to Austin Crinklaw
+
+    if not args["mhc_name"]:
+        generate_problem_table(
+            level="error",
+            rule_name=incorrect_num_str,
+            value=args["mhc_name"],
+            instructions="Enter MHC molecule",
+        )
+
+    if args["mod_pos"] and not args["mod_type"]:
+        generate_problem_table(
+            level="error",
+            rule_name=incorrect_num_str,
+            value=args["mod_type"],
+            instructions="Provide modificiation type(s)",
+        )
+
+    if args["mod_type"] and not args["mod_pos"]:
+        generate_problem_table(
+            level="error",
+            rule_name=incorrect_num_str,
+            value=args["mod_pos"],
+            instructions="Provide modificiation position(s)",
+        )
+
+
+def null_input_check(args):
+    null_rule_name = "NullValueEntered"
+    null_strs = [
```

Comment:
This list should not include any duplicates.

Reply:
Resolved by commit 229f774